### ISSUE TITLE
[backend] Boost atomics build hangs in infinite loop

### ISSUE THREAD

**Issue by HolyShitMan** [llvm/llvm-project#52727]
**Description:**
If building boost atomics with `-Og` and `-mavx512vl` on intel skylake-avx512 architecture clang hangs while using 100% CPU. Running clang in gdb shows that DAGCombiner::Run() runs infinitely when building  `libs/atomic/src/find_address_sse41.cpp`.

If -mllvm -debug is appended to the specific clang call log repeatedly shows:
```
Combining: t6706: v8i16 = truncate t6705, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                                    
Creating constant: t6707: i64 = Constant<0>                                                                                                                                                                                                                                                                                                                                                                                               
Creating constant: t6708: i64 = Constant<4>                                                                                                                                                                                                                                                                                                                                                                                               
Creating new node: t6709: v8i16 = X86ISD::PACKSS t200, t201, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                 
 ... into: t6709: v8i16 = X86ISD::PACKSS t200, t201, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                                                                                                                                          
Legalizing: t6709: v8i16 = X86ISD::PACKSS t200, t201, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                        
Legal node: nothing to do                                                                                                                                                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                                                                                                                                          
Combining: t6709: v8i16 = X86ISD::PACKSS t200, t201, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                         
Creating new node: t6710: v8i32 = concat_vectors t200, t201, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                 
Creating new node: t6711: v8i16 = truncate t6710, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                            
 ... into: t6711: v8i16 = truncate t6710, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                                                                                          
Legalizing: t6711: v8i16 = truncate t6710, libs/atomic/src/find_address_sse41.cpp:67:32                                                                                                                                                                                                                                                                                                                                                   
Trying custom legalization
```

Following the "How to submit an LLVM bug report" on llvm.org as expected nothing crashed, but `llc find_address_sse41.bc` also hangs. You can find_address_sse41.bc attached to this bug report.

[find_address_sse41.zip](https://github.com/llvm/llvm-project/files/7720877/find_address_sse41.zip)

**CPU Architecture:**
Intel skylake-avx512

**Tested clang versions:** 
* clang 13.0.0
* clang 14.0.0  (4cd6dc5adb7ab44ee31832770cfa0d99fe1f205e)

**Tested boost versions:**
Boost 1.77.0
Boost 1.78.0

**Boost build commands to reproduce:**
```
$ ./bootstrap.sh 
$ ./b2 -q -d0 -j 24 --with-atomic toolset=clang-13 --user-config=user-config.jam variant=debug install
```
Boost user-config.jam
```
using clang : 13 : clang-13 : <compileflags>" -mavx512vl -Og " <linkflags>"" ;
```

Regards
Simon

**Comment by dannftk** [comment]
Hi, I am a gentoo user and observe the same issue with boost and qtgui, both happening with files, names of which partially consist of 'sse41'. I do not think it is a coincidence, something has been broken with compilation of sse41-related operations
Thank you

**Comment by dannftk** [comment]
StackTrace:
```
#0  0x00007f9c5a7c99d4 in llvm::SelectionDAG::ReplaceAllUsesWith(llvm::SDNode*, llvm::SDNode*) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#1  0x00007f9c5a628572 in llvm::SelectionDAG::Combine(llvm::CombineLevel, llvm::AAResults*, llvm::CodeGenOpt::Level) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#2  0x00007f9c5a80aa44 in llvm::SelectionDAGISel::CodeGenAndEmitDAG() () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#3  0x00007f9c5a809e2d in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#4  0x00007f9c5a807a8e in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#5  0x00007f9c5ba405e5 in ?? () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#6  0x00007f9c5a38b8af in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#7  0x00007f9c5a12f836 in llvm::FPPassManager::runOnFunction(llvm::Function&) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#8  0x00007f9c5a1359d3 in llvm::FPPassManager::runOnModule(llvm::Module&) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#9  0x00007f9c5a12feeb in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#10 0x00007f9c5e4fda0b in ?? () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#11 0x00007f9c5e4f8049 in clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#12 0x00007f9c5e7c9c7b in ?? () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#13 0x00007f9c5d6e32e4 in clang::ParseAST(clang::Sema&, bool, bool) () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#14 0x00007f9c5f01f03f in clang::FrontendAction::Execute() () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#15 0x00007f9c5efa2f16 in clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#16 0x00007f9c5f092e52 in clang::ExecuteCompilerInvocation(clang::CompilerInstance*) () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#17 0x00000000002157b3 in cc1_main(llvm::ArrayRef<char const*>, char const*, void*) ()
#18 0x00000000002136d3 in ?? ()
#19 0x00007f9c5ec7f0d2 in ?? () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#20 0x00007f9c59f138ce in llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) () from target:/usr/lib/llvm/13/bin/../lib64/libLLVM-13libcxx.so
#21 0x00007f9c5ec7eba8 in clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*, bool*) const
    () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#22 0x00007f9c5ec4f53c in clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const () from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#23 0x00007f9c5ec4f8f7 in clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::__1::pair<int, clang::driver::Command const*> >&) const ()
   from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#24 0x00007f9c5ec6513c in clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::__1::pair<int, clang::driver::Command const*> >&) ()
   from target:/usr/lib/llvm/13/bin/../lib64/libclang-cpp.so.13libcxx
#25 0x0000000000212b68 in main ()
```

**Comment by fhahn** [comment]
Could you provide the LLVM IR file before the backend is run? You can get that by passing `-emit-llvm` to your Clang invocation.

**Comment by fhahn** [comment]
Nevermind, 

> Could you provide the LLVM IR file before the backend is run? You can get that by passing `-emit-llvm` to your Clang invocation.

never mind, I saw the zip file already contains it...

**Comment by dannftk** [comment]
Guys, is there a workaround? Or we are supposed to downshift to clang-12?

**Comment by fhahn** [comment]
> Guys, is there a workaround? Or we are supposed to downshift to clang-12?

Does building with -O1/-O2/-O3 also hang?

**Comment by dannftk** [comment]
> > Guys, is there a workaround? Or we are supposed to downshift to clang-12?
> 
> Does building with -O1/-O2/-O3 also hang?

-O2 definitely hangs, I haven't exercised other modes

**Comment by HolyShitMan** [comment]
For me -O3 runs through

**Comment by dannftk** [comment]
@fhahn I am sorry for having confused you but I omitted that I have been compiling with -O1 rather -O2, therefore I could conclude that -O1 definitely hangs up. There is nothing to say about -O2

**Comment by dannftk** [comment]
For me -O3 does also hangs

**Comment by wudisheng** [comment]
FYI: This is still occurring with clang-13/14/15 x boost-1.77/1.78/1.79 when avx512 is turned on.

**Comment by dpronin** [comment]
I am not an expert in llvm and clang, but we could take a look at the difference between llvm-12 and llvm-13 and inspect possible changes made against avx512, couldn't we?

**Comment by davidbolvansky** [comment]
Cc @RKSimon @topperc

**Comment by fhahn** [comment]
It looks like this has been fixed by e8305c0b8f492272446d517e67d451f390d35dbe (by @RKSimon , thanks!)


Please re-open with a new bitcode reproducer if it still reproduces on your end.
